runtime/drivers/duckdb/duckdb.go

```diff
@@ -17,15 +19,21 @@ func init() {
 type driver struct{}
 
 func (d driver) Open(dsn string) (drivers.Connection, error) {
-	db, err := sqlx.Open("duckdb", dsn)
+	cfg, err := newConfig(dsn)
 	if err != nil {
 		return nil, err
 	}
 
-	// database/sql has a built-in connection pool, but DuckDB loads extensions on a per-connection basis.
-	// So we allow only one open connection at a time. In the future, we may instead consider using db.Conn()
-	// and building our own connection pool to work around DuckDB's idiosyncracies.
-	db.SetMaxOpenConns(1)
+	// database/sql has a built-in connection pool, but DuckDB loads extensions on a per-connection basis,
+	// which means we need to manually initialize each connection before it's used.
+	// database/sql doesn't give us that flexibility, so we implement our own (very simple) pool.
+
+	db, err := sqlx.Open("duckdb", cfg.DSN)
+	if err != nil {
+		return nil, err
+	}
+	// 1 extra for meta connection to be used for metadata queries like catalog and dry run queries
+	db.SetMaxOpenConns(cfg.PoolSize + 1)
```

Comment:
This should be `db.SetMaxOpenConns(cfg.PoolSize)` so we don't create more connections than expected. Instead, make the semaphore size:
```go
semSize := cfg.PoolSize - 1
if semSize < 1 {
  semSize = 1
}
```
In the case where `cfg.PoolSize` is 1, it means the meta calls may block a while, but that's fine.

Reply:
using the built in pool so semSize not needed anymore, for now removed metaConn altogether. Will probably handle that separately

Reply:
when using built in pool, we will have to create and keep `metaConn` explicitly as we don't have control over the pool